core/src/main/java/org/jboss/modcluster/ModClusterService.java

```diff
@@ -300,10 +300,12 @@ protected void establishJvmRoute(Engine engine) {
     public void add(Context context) {
         ModClusterLogger.LOGGER.addContext(context.getHost(), context);
 
-        if (this.include(context)) {
+        if (this.include(context) && this.established) {
             // Send ENABLE-APP if state is started
-            if (this.established && context.isStarted()) {
+            if (context.isStarted()) {
                 this.enable(context);
+            } else {
+                this.stop(context);
```

Comment:
This appears racy – if `.enable(context)` registers the context as available there might be requests routed to this context before the stop command is received. I am afraid there might not be a simple fix without changing the MCMP protocol, but I ll have a better look.

Reply:
Also, note to self - check that this isn't racy in tomcat.

Reply:
@rhusar did you have a chance to look more into this?